### PR TITLE
[Student Libraries] Exporter UI

### PR DESCRIPTION
**Pull request by epeach**
# Description

Part of the Student Libraries work.
When the user goes to 'Share' -> 'Share as Library' this is what they see.

Note:
Not i18n until wording confirmed.

Note: Can't submit with an empty description or with no checkboxes checked. Can't check a checkbox that doesn't have a comment.

Peek 2019-10-16 11-38

## Links

- [spec](https://docs.google.com/document/d/1VxMmdHdIeBWCXKofkjEy9MuWcrzeEW3Zt-Ir8pINYdA/edit#heading=h.9a8gdpezdgnm)

## Testing story

Manually tested.
# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 1]
```diff
@@ -67,33 +90,85 @@ class LibraryCreationDialog extends React.Component {
...
       >
-        {this.displayFunctions()}
-        <button type="button" onClick={this.publish}>
-          {i18n.publish()}
-        </button>
-      </BaseDialog>
+        <Body>
+          <PadAndCenter>
+            <div style={styles.libraryBoundary}>
+              <Heading1>Functions as a Library</Heading1>
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Maybe "Export functions as a library?" (also posted in #student-created-libraries where more of the UI review is happening)

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 2]
```diff
@@ -67,33 +90,85 @@ class LibraryCreationDialog extends React.Component {
...
+        <form onSubmit={this.publish}>
+          <textarea
+            required
+            name="description"
+            rows="2"
+            cols="200"
+            style={styles.textarea}
+            placeholder="Write a description of your library"
+          />
+          {this.state.selectedFunctionList.map(selectedFunction => {
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: When 'publish' is clicked, only the functions whose boxes are checked should be in "selectedFunctionsList."

Comment (jmkulwik) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: I didn't add this originally, but we'll probably need two lists of functions, the original functions parsed from the code, and the list of functions that have been checked.

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 3]
```diff
@@ -1,12 +1,34 @@
...
 import PropTypes from 'prop-types';
-import BaseDialog from '../../../templates/BaseDialog';
+import Dialog, {Body} from '@cdo/apps/templates/Dialog';
 import {connect} from 'react-redux';
 import {hideLibraryCreationDialog} from '../shareDialogRedux';
 import libraryParser from './libraryParser';
 import LibraryClientApi from './LibraryClientApi';
 import i18n from '@cdo/locale';
+import PadAndCenter from '../../../templates/teacherDashboard/PadAndCenter';
+import {Heading1, Heading2} from '../../../lib/ui/Headings';
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Might be good to use @cdo format for these two as well given the three levels of ../.

**Comment by jmkulwik** [comment]
Make sure you add tests!
Some already exist under /code-dot-org/apps/test/unit/code-studio/components/libraries/

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 2 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Whoops - forgot to only send the selected functions through. :woman_facepalming: Fixing.

I don't understand your second comment though. Can you elaborate?

Comment (jmkulwik) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Haha it was meant to clarify the first comment, but sounds like it wasn't necessary (or helpful lol).

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 1 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: :heavy_check_mark:

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 2 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: For the createLibrary - I swapped out the full list of functions to be just the selected ones. Can you confirm that's the correct usage? (Latest commit)

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 3 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: :heavy_check_mark:

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 4]
```diff
@@ -119,16 +126,17 @@ class LibraryCreationDialog extends React.Component {
...
+            let name = sourceFunction.functionName;
+            let comment = sourceFunction.comment;
             return (
               <div key={keyIndex++} style={styles.functionItem}>
                 <input
                   type="checkbox"
                   style={styles.largerCheckbox}
                   disabled={comment.length === 0}
                   onClick={this.validateInput}
+                  value={keyIndex}
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: What does this do? (and why is the default -1?)

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 5]
```diff
@@ -71,9 +71,16 @@ class LibraryCreationDialog extends React.Component {
...
+    [...formElements].forEach(element => {
+      if (element.type === 'checkbox' && element.checked) {
+        selectedFunctionList.push(this.state.sourceFunctionList[element.value]);
+      }
+    });
     let libraryJson = libraryParser.createLibraryJson(
       this.state.librarySource,
-      this.state.selectedFunctionList,
+      selectedFunctionList,
       this.state.libraryName
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Were you planning to add libraryDescription here?

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 2 above)
Comment (jmkulwik) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: 👍 Thank you!

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 4 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Good question! This was a workaround for handling functions that have the same name/comments. Since we can't rely on unique values to connected the selected checkbox to the list of functions, I've shifted to using indices. Because javascript .map goes over the array items in order, we can assign each item created an index so that we can then track that index back to the function it is meant to represent.

Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: It is initialized at `-1` because it increments on each item created, so it has to start at `-1` so that it starts zero indexed. I don't love this solution, but I couldn't really figure out another way to do while still using a map. I'm open to suggestions!

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 5 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Yes, still in progress.

**Comment by epeach** [comment]
@jmkulwik PTAL - while implementing tests I realized that there are something in the state that are probably best suited to not be part of state. I am tracking an work item now to fix that at a later point and add some more tests, but I don't think it's worth blocking this PR on it.

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 6]
```diff
@@ -67,37 +104,92 @@ class LibraryCreationDialog extends React.Component {
     );
   };
 
+  validateInput = () => {
+    // Check if any of the checkboxes are checked
+    // If this changes the publishable state, update
+    let isChecked = $("input[type='checkbox']").is(':checked');
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Also check for a library description.

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 7]
```diff
@@ -67,37 +104,92 @@ class LibraryCreationDialog extends React.Component {
...
+                  disabled={comment.length === 0}
+                  onClick={this.validateInput}
+                  value={keyIndex}
+                />
+                {name}
+                <br />
+                {comment.length === 0 && (
+                  <p style={styles.alert}>
+                    This function cannot be exported until you add a comment to
+                    it.
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: i18n

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 4 above)
Comment (jmkulwik) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: Don't love that it starts at -1?
You should be able to increment the counter at the end of .map instead of the beginning of .map. Then you can start it at 0.

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 8]
```diff
@@ -48,11 +72,24 @@ class LibraryCreationDialog extends React.Component {
...
 
   publish = () => {
+    let formElements = document.getElementById('selectFunction').elements;
+    let selectedFunctionList = [];
+    let libraryDescription = '';
+    [...formElements].forEach(element => {
+      if (element.type === 'checkbox' && element.checked) {
+        selectedFunctionList.push(this.state.sourceFunctionList[element.value]);
+      }
+      if (element.type === 'textarea') {
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx:81]: NIT: (I don't feel too strongly about this) This seems like it would be more maintainable as a check by class name rather than type. i.e. it's not too unlikely that we'd add another textarea somewhere.

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 7 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx]: I was planning to hold off on i18n until we had a better idea if these were the final phrasings. Do you think we're there?

**Comment by epeach** [comment]
@jmkulwik PTAL - thanks!

File: apps/test/unit/code-studio/components/libraries/LibraryCreationDialogTest.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+      loadingFinished: true,
+      sourceFunctionList: libraryParser.getFunctions(LIBRARY_SOURCE)
+    });
+
+    wrapper.update();
+
+    expect(wrapper.find(CHECKBOX_SELECTOR).last()).to.be.disabled();
+  });
+
+  it('displays loading while in the loading state', () => {
```
Comment (jmkulwik) [inline apps/test/unit/code-studio/components/libraries/LibraryCreationDialogTest.js:81]: Awesome

File: apps/test/unit/code-studio/components/libraries/libraryParserTest.js [hunk 10]
```diff
@@ -130,10 +175,16 @@ describe('Library parser', () => {
 
       let expectedFunctions = `${firstFunctionName}: ${firstFunctionName},${secondFunctionName}: ${secondFunctionName}`;
```
Comment (jmkulwik) [inline apps/test/unit/code-studio/components/libraries/libraryParserTest.js:176]: Maybe add a test with a non-empty description

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx [hunk 11]
```diff
@@ -67,37 +103,97 @@ class LibraryCreationDialog extends React.Component {
...
+    });
+    if (isChecked !== this.state.canPublish) {
+      this.setState({canPublish: isChecked});
+    }
+  };
+
   displayFunctions = () => {
     if (!this.state.loadingFinished) {
-      return <div>Loading...</div>;
+      return <div id="loading">Loading...</div>;
```
Comment (jmkulwik) [inline apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx:123]: If you get the chance, it would be nice to make this a spinner instead of `Loading...`

**Review by jmkulwik — APPROVED** [review]

File: apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx (hunk 11 above)
Comment (epeach) [reply apps/src/code-studio/components/Libraries/LibraryCreationDialog.jsx:123]: https://codedotorg.atlassian.net/browse/STAR-775

File: apps/test/unit/code-studio/components/libraries/libraryParserTest.js (hunk 10 above)
Comment (epeach) [reply apps/test/unit/code-studio/components/libraries/libraryParserTest.js:176]: Added to: https://codedotorg.atlassian.net/browse/STAR-765
